pfrl/agents/hrl/hrl_controllers.py

```diff
@@ -46,29 +48,39 @@ def __init__(
         self.is_low_level = is_low_level
         self.minibatch_size = minibatch_size
         self.add_entropy = add_entropy
-        # create td3 agent
-        self.device = torch.device(f'cuda:{gpu}')
+
+        # create agent
         if self.add_entropy:
             def squashed_diagonal_gaussian_head(x):
-                mean, log_scale = torch.chunk(x, 2, dim=-1)
+                """
+                taken from the SAC code.
+                """
+                assert x.shape[-1] == action_dim * 2
+                mean, log_scale = torch.chunk(x, 2, dim=1)
                 log_scale = torch.clamp(log_scale, -20.0, 2.0)
                 var = torch.exp(log_scale * 2)
                 base_distribution = distributions.Independent(
                     distributions.Normal(loc=mean, scale=torch.sqrt(var)), 1
                 )
-                return base_distribution
+                # cache_size=1 is required for numerical stability
+                return distributions.transformed_distribution.TransformedDistribution(
+                    base_distribution, [distributions.transforms.TanhTransform(cache_size=1)]
+                )
 
+            # SAC policy definition:
             policy = nn.Sequential(
-                nn.Linear(state_dim + goal_dim, 300),
+                nn.Linear(state_dim + goal_dim, 256),
                 nn.ReLU(),
-                nn.Linear(300, 300),
+                nn.Linear(256, 256),
                 nn.ReLU(),
-                nn.Linear(300, action_dim * 2),
-                nn.Tanh(),
-                ConstantsMult(torch.cat((torch.tensor(self.scale), torch.ones(self.scale.size))).float().to(self.device)),
-                # pfrl.policies.DeterministicHead(),
+                nn.Linear(256, action_dim * 2),
                 Lambda(squashed_diagonal_gaussian_head),
                 )
+
+            torch.nn.init.xavier_uniform_(policy[0].weight)
+            torch.nn.init.xavier_uniform_(policy[2].weight)
+            torch.nn.init.xavier_uniform_(policy[4].weight)
+
```

Comment:
Is it standard to include Xavier Uniform Initialization in SAC? I do understand why it requires it somehow but what happens if we don't have them?

Reply:
I believe it is standard, from seeing the PFRL implementation of SAC.

Reply:
k